dynacred/src/group/groupContractCollection.ts

```diff
@@ -0,0 +1,341 @@
+            if (isAppended && this.isAccepted(groupContract)) {
+                this.currentGroupContract = groupContract;
+            }
+
+            return isAppended;
+        } catch (e) {
+            throw e;
+        }
+    }
+
+    /**
+     * Append a group contract to this collection
+     *
+     * @param groupContract
+     * @param keepOnly if true do not make the group contract the current one
+     */
+    append(groupContract: GroupContract, keepOnly: boolean = false) {
+        try {
+            // only proceed if the the groupContract is sound
+            const parents = this.getParent(groupContract);
+            if (parents.length) {
+                if (!groupContract.verify(...this.getParent(groupContract))) {
+                    throw new TypeError("The group contract verification failed.");
+                }
+            } else {
+                if (!groupContract.verify(undefined)) {
+                    throw new TypeError("The group contract verification failed.");
+                }
+            }
+
+            // check if the id is not already there
+            const existing: GroupContract[] = [];
+            this._collection.forEach((gd: GroupContract) => {
+                if (gd.id === groupContract.id) {
+                    existing.push(gd);
+                }
+            });
+
+            if (existing.length) {
+                groupContract.mergeSignoffs(existing[0]);
+                this._collection.set(groupContract.id, groupContract);
+            } else {
+                // there is a new proposed group contract; therefore, erase the current proposed group contract
+                this.removeProposedGroupContract();
+
+                this._collection.set(groupContract.id, groupContract);
+            }
+
+            // if groupContract is accepted; therefore, it becomes the current group contract
+            const numbPredecessor = groupContract.groupDefinition.predecessor.length;
+            if (!keepOnly && (numbPredecessor === 0 || (numbPredecessor > 0 && this.isAccepted(groupContract)))) {
+                this.currentGroupContract = groupContract;
+            }
+        } catch (e) {
+            throw e;
+        }
+    }
+
+    /**
+     *
+     * @param groupContract
+     * @returns {boolean} true if the collection has the group contract, otherwise false
+     */
+    has(groupContract: GroupContract): boolean {
+        return this._collection.has(groupContract.id);
+    }
+
+    /**
+     * Returns group contract by id
+     *
+     * @param id
+     * @returns {GroupContract}
+     */
+    get(id: string): GroupContract {
+        return this._collection.get(id);
+    }
+
+    getCurrentGroupContract(): GroupContract {
+        return this.currentGroupContract;
+    }
+
+    getProposedGroupContract(): GroupContract {
+        try {
+            for (const gc of Array.from(this._collection.values())) {
```

Comment:
I have a hard finding a way to do it only using `reduce`. Can you give me some hint, please?

Reply:
```javascript
return Array.from(this._collection.values())
    .filter(gc => gc.groupDefinition.predecessor.length > 0 && !this.isAccepted(gc))
    .reduce((_, cur) => cur, undefined)
```

In this case, it will not return the first matching element but the last one, maybe that's not intended. If so, you can use the accumulator argument for `reduce` and if it isn't `undefined`, such as
```javascript
.reduce((acc, cur) => acc === undefined ? cur : undefined, undefined)
```